pypeit/core/wavecal/autoid.py

```diff
@@ -0,0 +1,799 @@
+
+    if debug:
+        match_idx = best_dict['midx']
+        for kk in match_idx.keys():
+            uni, counts = np.unique(match_idx[kk]['matches'], return_counts=True)
+            msgs.info('kk={}, {}, {}, {}'.format(kk, uni, counts, np.sum(counts)))
+
+    # Write scores
+    #out_dict = best_dict['scores']
+    #jdict = ltu.jsonify(out_dict)
+    #ltu.savejson(pargs.outroot+'.scores', jdict, easy_to_read=True, overwrite=True)
+
+    # Write IDs
+    if outroot is not None:
+        out_dict = dict(pix=cut_tcent, IDs=best_dict['IDs'])
+        jdict = ltu.jsonify(out_dict)
+        ltu.savejson(outroot+'.json', jdict, easy_to_read=True, overwrite=True)
+        msgs.info("Wrote: {:s}".format(outroot+'.json'))
+
+    # Plot
+    if outroot is not None:
+        tmp_list = vstack([line_lists,unknwns])
+        qa.match_qa(spec, cut_tcent, tmp_list, best_dict['IDs'], best_dict['scores'], outroot+'.pdf')
+        msgs.info("Wrote: {:s}".format(outroot+'.pdf'))
+
+    # Fit
+    final_fit = None
+    if do_fit:
+        '''
+        # Read in Full NIST Tables
+        full_NIST = waveio.load_line_lists(lines, NIST=True)
+        # KLUDGE!!!!!
+        keep = full_NIST['wave'] > 8800.
+        pdb.set_trace()
+        line_lists = vstack([line_lists, full_NIST[keep]])
+        '''
+        #
+        NIST_lines = line_lists['NIST'] > 0
+        ifit = np.where(best_dict['mask'])[0]
+        if outroot is not None:
+            plot_fil = outroot+'_fit.pdf'
+        else:
+            plot_fil = None
+        # Purge UNKNOWNS from ifit
+        imsk = np.array([True]*len(ifit))
+        for kk, idwv in enumerate(np.array(best_dict['IDs'])[ifit]):
+            if np.min(np.abs(line_lists['wave'][NIST_lines]-idwv)) > 0.01:
+                imsk[kk] = False
+        ifit = ifit[imsk]
+        # Allow for weaker lines in the fit
+        all_tcent, weak_cut_tcent, icut = utils.arc_lines_from_spec(spec, min_ampl=lowest_ampl)
+        add_weak = []
+        for weak in weak_cut_tcent:
+            if np.min(np.abs(cut_tcent-weak)) > 5.:
+                add_weak += [weak]
+        if len(add_weak) > 0:
+            cut_tcent = np.concatenate([cut_tcent, np.array(add_weak)])
+        # Fit
+        final_fit = fitting.iterative_fitting(spec, cut_tcent, ifit,
+                                              np.array(best_dict['IDs'])[ifit], line_lists[NIST_lines],
+                                              disp, plot_fil=plot_fil, verbose=verbose, aparm=fit_parm)
+        if plot_fil is not None:
+            print("Wrote: {:s}".format(plot_fil))
+
+    # Return
+    return best_dict, final_fit
+
+
+def general(spec, lines, ok_mask=None, min_ampl=300., islinelist=False,
+            outroot=None, debug=False, do_fit=True, verbose=False,
+            fit_parm=None, min_nmatch=0, lowest_ampl=200.,
+            binw=None, bind=None, nstore=10, use_unknowns=False):
+    """
+    Parameters
+    ----------
+    spec
```

Comment:
This one I will 'demand' we document right.

Reply:
done - I should have remembered to do this one!